File: tests/domains/video/test_zoom.py
import pytest

from domains import Runtime
from domains.video.probe import probe_video
from domains.video.zoom import zoom_expression, zoom_video
from tests.helpers import unwrap

pytestmark = pytest.mark.ffmpeg


def test_zoom_expression() -> None:
    assert zoom_expression("punch", 1.5, 1.0, 2.0) == (
        "if(between(t\\,1.000\\,2.000)\\,1.5000\\,1)"
    )
    assert "(1+(1.5000-1)*" in zoom_expression("in", 1.5, 0.0, 1.0)
    assert "(1.5000-(1.5000-1)*" in zoom_expression("out", 1.5, 0.0, 1.0)


@pytest.mark.parametrize("mode", ["punch", "in", "out"])
def test_zoom_video(runtime: Runtime, sample_video: str, mode: str) -> None:
    result = unwrap(zoom_video(runtime, sample_video, 0.5, 2.0, mode=mode, focus_y=0.3))  # type: ignore[arg-type]
    assert result["output"] == f"sample_zoom_{mode}.mp4"
    probed = unwrap(probe_video(runtime, result["output"]))
    assert probed["info"]["width"] == 320
    assert probed["info"]["height"] == 240
    assert abs(probed["info"]["duration"] - 3.0) < 0.3


def test_zoom_video_invalid(runtime: Runtime, sample_video: str) -> None:
    assert zoom_video(runtime, sample_video, 2.0, 1.0).get("code") == "invalid_argument"
    assert zoom_video(runtime, sample_video, 0.0, 1.0, zoom=10).get("code") == "invalid_argument"
    assert zoom_video(runtime, sample_video, 0.0, 1.0, focus_x=2).get("code") == (
        "invalid_argument"
    )
    assert zoom_video(runtime, sample_video, 50.0, 51.0).get("code") == "invalid_argument"


def test_zoom_video_missing(runtime: Runtime) -> None:
    assert zoom_video(runtime, "ghost.mp4", 0.0, 1.0).get("code") == "not_found"

File: version-in-rust/tests/domains/video/test_zoom.rs
use mcp_tools::core::errors::ErrorCode;
use mcp_tools::domains::video::probe::probe_video;
use mcp_tools::domains::video::zoom::{zoom_expression, zoom_video, ZoomMode};

use crate::conftest::{runtime, sample_video};
use crate::helpers::unwrap;
use crate::skip_without_ffmpeg;

#[test]
fn test_zoom_expression() {
    assert_eq!(
        zoom_expression(ZoomMode::Punch, 1.5, 1.0, 2.0),
        "if(between(t\\,1.000\\,2.000)\\,1.5000\\,1)"
    );
    assert!(zoom_expression(ZoomMode::In, 1.5, 0.0, 1.0).contains("(1+(1.5000-1)*"));
    assert!(zoom_expression(ZoomMode::Out, 1.5, 0.0, 1.0).contains("(1.5000-(1.5000-1)*"));
}

#[test]
fn test_zoom_video() {
    skip_without_ffmpeg!();
    for mode in [ZoomMode::Punch, ZoomMode::In, ZoomMode::Out] {
        let rt = runtime();
        let sample = sample_video(&rt.runtime.workspace);
        let result = unwrap(zoom_video(
            &rt.runtime,
            &sample,
            0.5,
            2.0,
            mode,
            1.3,
            0.5,
            0.3,
            false,
        ));
        assert_eq!(result.output, format!("sample_zoom_{}.mp4", mode.as_str()));
        let probed = probe_video(&rt.runtime, &result.output).unwrap();
        assert_eq!(probed.info.width, Some(320));
        assert_eq!(probed.info.height, Some(240));
        assert!((probed.info.duration - 3.0).abs() < 0.3);
    }
}

#[test]
fn test_zoom_video_invalid() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let punch = ZoomMode::Punch;
    let cases: [(f64, f64, f64, f64); 4] = [
        (2.0, 1.0, 1.3, 0.5),
        (0.0, 1.0, 10.0, 0.5),
        (0.0, 1.0, 1.3, 2.0),
        (50.0, 51.0, 1.3, 0.5),
    ];
    for (start, end, zoom, focus_x) in cases {
        let error = zoom_video(
            &rt.runtime,
            &sample,
            start,
            end,
            punch,
            zoom,
            focus_x,
            0.5,
            false,
        )
        .unwrap_err();
        assert_eq!(error.code, ErrorCode::InvalidArgument);
    }
}

#[test]
fn test_zoom_video_missing() {
    let rt = runtime();
    let error = zoom_video(
        &rt.runtime,
        "ghost.mp4",
        0.0,
        1.0,
        ZoomMode::Punch,
        1.3,
        0.5,
        0.5,
        false,
    )
    .unwrap_err();
    assert_eq!(error.code, ErrorCode::NotFound);
}
